File: 0075_sort_colors/sort_colors.js
/**
 * Time:    O(n)
 * Space:   O(1)
 * @param   {number[]} nums
 * @return
 */
const sortColors = nums => {
  let p0, c, p2
  p0 = c = 0
  p2 = nums.length - 1
  while (c <= p2) {
    if (nums[c] === 0) [nums[c++], nums[p0++]] = [nums[p0], nums[c]]
    else if (nums[c] === 2) [nums[c], nums[p2--]] = [nums[p2], nums[c]]
    else c++
  }
}

File: 0075_sort_colors/sort_colors.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(1)
 */
void sortColors(vector<int> &nums)
{
    int c = 0, p0 = 0, p2 = nums.size() - 1;
    while (c <= p2)
    {
        if (nums[c] == 0)
            swap(nums[c++], nums[p0++]);
        else if (nums[c] == 2)
            swap(nums[c], nums[p2--]);
        else
            c++;
    }
}
